### PR TITLE
Select manual filtering.

### PR DESCRIPTION
**Pull request by atmgrifter00**
# Pull Request

## 🤨 Rationale

- #1866

## 👩‍💻 Implementation

This PR surfaces the necessary APIs for clients to be able to provide their own manual filtering of options in the `Select`. Additionally, clients can now set a `loading-visible` attribute on the `Select` to `true` to show a visual in the dropdown to indicate that it is in a loading state.

Not in this PR:
- Examples that actually exercise the expected client workflow in which they would apply their own custom filtering for options. Instead, there are simply unit tests that validate the necessary notifications for clients to use (in addition to tests regarding the loading visual in the dropdown). The Angular and/or Blazor integration will provide a more complete example of this feature.

## 🧪 Testing

Unit tests. Chromatic tests.

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/select/tests/select.spec.ts [hunk 1]
```diff
@@ -833,412 +845,468 @@ describe('Select', () => {
...
         let pageObject: SelectPageObject;
 
-        beforeEach(async () => {
-            ({ element, connect, disconnect } = await setup());
-            element.style.width = '200px';
-            element.filterMode = FilterMode.standard;
-            await connect();
-            pageObject = new SelectPageObject(element);
-        });
+        describe('with filterMode = standard', () => {
```
Comment (atmgrifter00) [inline packages/nimble-components/src/select/tests/select.spec.ts:848]: All changes from here until line 1271 are indentation. Differ is a little confused in places.

**Comment by atmgrifter00** [comment]
@m-akinc you mind buddying this?

File: packages/nimble-components/src/select/template.ts [hunk 2]
```diff
@@ -151,11 +153,19 @@ SelectOptions
...
                         ></slot>
                     </div>
-                    ${when(x => (x.filterMode !== FilterMode.none && x.filteredOptions.length === 0), html<Select>`
+                    ${when(x => (x.filterMode !== FilterMode.none && x.filteredOptions.length === 0 && !x.loadingVisible), html<Select>`
                         <span class="no-results-label ${x => x.positionAttribute}">
                             ${x => filterNoResultsLabel.getValueFor(x)}
                         </span>
                     `)}
+                    ${when(x => x.loadingVisible, html<Select>`
+                        <div class="loading-container" @click="${(x, c) => x.ignoreClickHandler(c.event as MouseEvent)}">
```
Comment (m-akinc) [inline packages/nimble-components/src/select/template.ts:162]: Do we need `ignoreClickHandler` here?

File: packages/nimble-components/src/select/styles.ts [hunk 3]
```diff
@@ -151,6 +151,28 @@ export const styles = css`
...
+
+    .loading-container.above {
+        margin-top: ${mediumPadding};
+        margin-bottom: 0;
+    }
+
+    .loading-container.below {
+        margin-top: 0;
+        margin-bottom: ${mediumPadding};
+    }
```
Comment (m-akinc) [inline packages/nimble-components/src/select/styles.ts:171]: I don't see the loading container ever getting a class of `above` or `below`.

File: packages/nimble-components/src/select/tests/select.spec.ts [hunk 4]
```diff
@@ -831,413 +843,462 @@ describe('Select', () => {
...
+                await pageObject.openAndSetFilterText('o');
+                const filteredOptions = pageObject.getFilteredOptions();
+                expect(filteredOptions.length).toBe(6);
+            });
+
+            it('emits filter-input event when filter text is entered', async () => {
+                const filterInputEvent = jasmine.createSpy();
+                element.addEventListener('filter-input', filterInputEvent);
+                await pageObject.openAndSetFilterText('o');
+                expect(filterInputEvent).toHaveBeenCalledTimes(1);
```
Comment (m-akinc) [inline packages/nimble-components/src/select/tests/select.spec.ts:1286]: Validate event `filterText` property?

File: packages/nimble-components/src/select/tests/select.spec.ts [hunk 5]
```diff
@@ -831,413 +843,462 @@ describe('Select', () => {
...
+            });
 
-            expect(option.visuallyHidden).toBeFalse();
+            it('emits filter-input event with empty filterText when dropdown is closed', async () => {
```
Comment (m-akinc) [inline packages/nimble-components/src/select/tests/select.spec.ts:1289]: Add test case that event is emitted when opening dropdown?

File: packages/nimble-components/src/select/tests/select.spec.ts [hunk 6]
```diff
@@ -831,413 +843,462 @@ describe('Select', () => {
...
-            expect(option.visuallyHidden).toBeTrue();
+                expect(option.visuallyHidden).toBeFalse();
+            });
         });
 
-        it('option that is not hidden but visuallyHidden, if removed from DOM, and then re-added to DOM while dropdown closed, is visible in dropdown when opened', async () => {
-            const option = element.options[1] as ListOption;
-            await pageObject.openAndSetFilterText('Three');
-            expect(option.visuallyHidden).toBeTrue();
+        describe('with filterMode = manual', () => {
```
Comment (m-akinc) [inline packages/nimble-components/src/select/tests/select.spec.ts:1263]: Is there anything interesting to test about the `filterMode` being changed?

File: packages/nimble-components/src/select/tests/select.spec.ts [hunk 7]
```diff
@@ -1878,6 +1939,20 @@ describe('Select', () => {
...
             });
+
+            it('if all results are hidden, and loadingVisible is true, no results label is not shown', async () => {
+                element.loadingVisible = true;
+                await waitForUpdatesAsync();
+                await pageObject.openAndSetFilterText('abc');
+                expect(pageObject.isNoResultsLabelVisible()).toBeFalse();
+            });
+
+            it('does not automatically filter options when filterMode is manual', async () => {
```
Comment (m-akinc) [inline packages/nimble-components/src/select/tests/select.spec.ts:1950]: You have another test with the same name earlier in the file. Is it a duplicate?

File: packages/nimble-components/src/select/styles.ts
Comment (m-akinc) [inline packages/nimble-components/src/select/styles.ts:1]: There's a slight spacing inconsistency when the loading indicator is shown alongside options. The loading indicator looks misaligned vertically, because it has different spacing between the text and the edge of the dropdown vs between the text and the adjacent option's text. This might be related to the margin styling I commented on earlier.
image
image

Comment (m-akinc) [inline packages/nimble-components/src/select/styles.ts:1]: Looks like we need to prevent the loading indicator from shrinking vertically when there isn't enough space to show it and all the options:
image

**Review by m-akinc — APPROVED** [review]

File: packages/nimble-components/src/label-provider/core/label-token-defaults.ts [hunk 8]
```diff
@@ -10,5 +10,6 @@ export const coreLabelDefaults: { readonly [key in TokenName]: string } = {
     popupIconWarningLabel: 'Warning',
     popupIconInformationLabel: 'Information',
     filterSearchLabel: 'Search',
-    filterNoResultsLabel: 'No items found'
+    filterNoResultsLabel: 'No items found',
+    loadingLabel: 'Loading'
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply packages/nimble-components/src/label-provider/core/label-token-defaults.ts:14]: @fredvisser? @jattasNI? I'm indifferent.

File: packages/nimble-components/src/select/template.ts (hunk 2 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/template.ts:162]: It's here so the dropdown isn't dismissed if a user clicks that visual. This avoids closing on it on misclicks.

File: packages/nimble-components/src/select/tests/select.spec.ts (hunk 6 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/tests/select.spec.ts:1263]: I don't believe so. There's no lingering state to be concerned about in either mode.

File: packages/nimble-components/src/select/tests/select.spec.ts (hunk 5 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/tests/select.spec.ts:1289]: We don't emit event on dropdown open. In the `openChanged` implementation you will see an early return if the dropdown is open.

File: packages/nimble-components/src/select/tests/select.spec.ts (hunk 7 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/tests/select.spec.ts:1950]: It's the same test, just done, additionally, when there are groups present, as that exercises a slightly different code path.

File: packages/nimble-components/src/select/styles.ts (hunk 3 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/styles.ts]: Yep. This, and the fact that the styles were wrong as-is, was the reason for the vertical padding issues you mentioned.

File: packages/nimble-components/src/label-provider/core/label-token-defaults.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rajsite) [reply packages/nimble-components/src/label-provider/core/label-token-defaults.ts:14]: Indifferent, other than if it is used we use the single character unicode ellipses instead of three periods.

```suggestion
    loadingLabel: 'Loading…'
```

[the remark this answers is not in the document]
Comment (jattasNI) [reply packages/nimble-components/src/label-provider/core/label-token-defaults.ts:14]: I'd vote to include the ellipsis.

I did a very brief search for guidance in other design systems and found that [Adobe Spectrum suggests](https://spectrum.adobe.com/page/grammar-and-mechanics/#Ellipsis) to "use an ellipsis at the end of the text to communicate the progress of something that is in process, such as a waiting or loading state." So at least we won't be the only ones taking this approach.

File: packages/nimble-components/src/select/tests/select.spec.ts (hunk 7 above)
Comment (jattasNI) [reply packages/nimble-components/src/select/tests/select.spec.ts:1956]: No strong opinions about how it's achieved, but they should have different resolved names (the name that appends all the nested `describe` descriptions) so that you can tell which test is failing from a log file.

File: packages/storybook/src/nimble/select/select.mdx [hunk 9]
```diff
@@ -52,6 +52,23 @@ the end of the prompt.
 
 <Canvas of={selectStories.placeholder} />
 
+### Dynamic Options
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.mdx:55]: I've started standardizing on sentence casing for mdx headings
```suggestion
### Dynamic options
```

File: packages/storybook/src/nimble/select/select.mdx [hunk 10]
```diff
@@ -52,6 +52,23 @@ the end of the prompt.
...
+listen for changes to the filter input. When accompanied with the `filterMode` set
+to `manual`, clients can perform their own filtering and provide the set of options
+to the `Select` dynamically.
+
+Typical steps for dynamic options are as follows:
+1. Set the `filterMode` to `manual` on the `Select`.
+1. Listen for the `filter-input` event on the `Select`.
+1. In the event handler, retrieve the filter text from the event and populate the DOM
+with the set of options that match the filter text by your own logic.
+    - It is important that one of the options in the DOM is the currently selected option.
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.mdx:70]: 1. generally prefer attribute names
2. make docs "timeless"
3. use either `nimble-select` or just "select"

```suggestion
There are scenarios that require not all possible options to be available
initially, such as when the set of all options is too large to be loaded at once. 
In such cases, users are expected to provide filter text to narrow down the list 
of options. The select offers the `filter-input` event to allow clients to
listen for changes to the filter input. When accompanied with the setting `filter-mode`
to `manual`, clients can perform their own filtering and provide the set of options
to the select dynamically.

Typical steps for dynamic options are as follows:
1. Set the `filter-mode` to `manual` on the select.
1. Listen for the `filter-input` event on the select.
1. In the event handler, retrieve the filter text from the event and populate the DOM
with the set of options that match the filter text by your own logic.
    - It is important that one of the options in the DOM is the currently selected option.
```

File: packages/storybook/src/nimble/select/select.mdx [hunk 11]
```diff
@@ -52,6 +52,23 @@ the end of the prompt.
...
+
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.mdx:71]: In the HLD we have a bunch of useful guidance about how clients should do filtering (copied below). Can you summarize the relevant points and add them here? I think setting `loading-visible` is particularly important.

----

IMPORTANT: When using the manual filter mode, clients are responsible for the following:

- Either adding/removing options to/from the DOM that do or do not match the filter and/or marking existing options as hidden when they don't match.
- The currently selected option is included in the set of options placed in the DOM regardless of match against filter (and be marked as hidden if it doesn't match the filter)
- If a placeholder option exists for the select, it should never be filtered out.

Notes:

- When the loading-visible attribute is set, we will display localizable text at the bottom of the dropdown (defaulting to "Loading"), along with the spinner icon.
- The Select will not peform any debouncing as the user types into the filter input. It is expected that clients can perform any debouncing that is needed easily at the app level.
- The filter-input event will be emitted for all filter modes (except none)

Groups

When using the filterMode: manual option along with [groups](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/select/specs/option-groups-hld.md), clients will also need to determine the appropriate matching behavior against the ListOptionGroup elements. Our guidance will be that filter text should be able to be matched against any text in the group label, and when that matches, all options within that group should be visible. Conversely, the select will automatically hide groups with no items in them. This means clients only need to conditionally remove items, not groups (though it's fine if they manually remove empty groups too).

File: packages/nimble-components/src/select/tests/select.spec.ts (hunk 7 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/tests/select.spec.ts:1956]: Yep. There is a describe block for "groups" that disambiguates them.

File: packages/storybook/src/nimble/select/select-opened-matrix.stories.ts [hunk 12]
```diff
@@ -46,9 +51,13 @@ interface SelectMatrixStoryOptions {
 
 // prettier-ignore
 const component = ({
-    positionState, filterMode, placeholder, grouped, optionsOutsideGroup, slottedLabel
+    positionState, filterMode, loadingVisible, placeholder, grouped, optionsOutsideGroup, slottedLabel
 }: SelectMatrixStoryOptions): ViewTemplate => html`
-    <${selectTag} open position="${() => positionState[0]}" style="${() => positionState[1]}" filter-mode="${() => filterMode}" style="width: 250px;">
+    <${selectTag} open 
+        position="${() => positionState[0]}" 
+        style="${() => positionState[1]} width: 250px; ${isChromatic() ? '--ni-private-spinner-animation-play-state:paused' : ''}"
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply packages/storybook/src/nimble/select/select-opened-matrix.stories.ts:58]: From what I can see it's working. Successive updates at least aren't triggering differences (that are unexpected).

File: packages/storybook/src/nimble/select/select-opened-matrix.stories.ts [hunk 13]
```diff
@@ -7,6 +7,7 @@ import { FilterMode } from '../../../../nimble-components/src/select/types';
 import { createFixedThemeStory } from '../../utilities/storybook';
 import { sharedMatrixParameters } from '../../utilities/matrix';
 import { backgroundStates } from '../../utilities/states';
+import { isChromatic } from '../../utilities/isChromatic';
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select-opened-matrix.stories.ts:10]: Reminder to take some care when Chromatic diffs. It looks like [the current build](https://www.chromatic.com/test?appId=60e89457a987cf003efc0a5b&id=6669161f7a2ba05516df3199) is exhibiting #1862 where sub menus are positioned far away from their parent, so we don't want to set that as the new baseline. My best advice is just to trigger a new build.

File: packages/storybook/src/nimble/select/select.stories.ts [hunk 14]
```diff
@@ -99,14 +101,18 @@ const optionSets = {
...
 
 It is recommended that if the select has 15 or fewer options that you use the \`none\` setting for the \`filter-mode\`.
 `;
 const clearableDescription = `
 When the \`clearable\` attribute is set, a clear button will be displayed in the select when a value is selected. Clicking the clear button will clear the selected value and display the placeholder text, if available, or will result in a blank display.
 `;
 
+const loadingVisibleDescription = `
+When the \`loading-visible\` attribute is set, a loading spinner will be displayed in the dropdown of the select along with localizable text that defaults to "Loading". This is useful when the select is loading its options dynamically.
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.stories.ts:113]:
```suggestion
When the \`loading-visible\` attribute is set, a loading spinner will be displayed in the dropdown of the select along with localizable text that defaults to "Loading…". This is useful when the select is loading its options dynamically.
```

File: packages/storybook/src/nimble/select/select.stories.ts [hunk 15]
```diff
@@ -235,6 +247,11 @@ export const select: Meta<SelectArgs> = {
             description: 'Emitted when the user changes the selected option.',
             table: { category: apiCategory.events },
             control: false
+        },
+        filterInput: {
+            description: 'Emitted when the user types in the filter input.',
```
Comment (jattasNI) [inline packages/storybook/src/nimble/select/select.stories.ts:252]:
```suggestion
            name: 'filter-input',
            description: 'Emitted when the user types in the filter input.',
```

File: packages/nimble-components/src/select/index.ts [hunk 16]
```diff
@@ -507,22 +510,30 @@ export class Select
...
+            ) {
+                activeOptionIndex = this.options.indexOf(enabledOptions[0]!);
+            } else if (enabledOptions.length === 0) {
+                activeOptionIndex = -1;
+            }
+            this.setActiveOption(activeOptionIndex);
+        }
+
+        if (this.filterMode !== FilterMode.none) {
+            this.emitFilterInputEvent();
```
Comment (jattasNI) [inline packages/nimble-components/src/select/index.ts:535]: In [the storybook build](https://60e89457a987cf003efc0a5b-cllvgqkkug.chromatic.com/?path=/story/components-select--select&args=filterMode:standard) in the Actions tab, I'm seeing the event be emitted twice for each key press. I know Storybook used to have bugs that would cause this but I haven't seen them recently. Can you check if that's a real issue or a Storybook issue? 

I'm also seeing `change` be fired twice so I suspect Storybook, but worth confirming.

File: packages/nimble-components/src/select/index.ts [hunk 17]
```diff
@@ -853,6 +864,9 @@ export class Select
             this.filterInput.value = '';
         }
 
+        if (this.filterMode !== FilterMode.none) {
+            this.emitFilterInputEvent();
+        }
```
Comment (rajsite) [inline packages/nimble-components/src/select/index.ts:869]: Can this check be merged with the check above it? Or is there some subtle difference in these conditions?
```ts
if (this.filterInput) {...}
if (this.filterMode !== FilterMode.none) {...}
```
Aren't those the same check?

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/select/index.ts (hunk 16 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/select/index.ts:535]: Fixed.

File: packages/storybook/src/nimble/select/select.stories.ts [hunk 18]
```diff
@@ -245,10 +242,26 @@ export const select: Meta<SelectArgs> = {
...
-        clearable: false
+        clearable: false,
+        loadingVisible: false
     }
 };
 
+export default metadata;
+
+export const select: StoryObj<SelectArgs> = {
+    ...metadata,
```
Comment (rajsite) [inline packages/storybook/src/nimble/select/select.stories.ts:253]: I don't think we have had to do this before. And it runs contrary to my understanding of how metadata works in storybook. i.e. should already expect that metadata is shared between each story. Seems fishy

Comment (atmgrifter00) [reply packages/storybook/src/nimble/select/select.stories.ts:253]: I'll try without it.

Comment (atmgrifter00) [reply packages/storybook/src/nimble/select/select.stories.ts]: Removed.

File: packages/nimble-components/src/select/styles.ts [hunk 19]
```diff
@@ -151,6 +151,37 @@ export const styles = css`
         align-items: center;
         padding: ${smallPadding} ${mediumPadding};
     }
+
+    .loading-container {
+        padding-left: ${mediumPadding};
+        padding-right: ${mediumPadding};
+        display: flex;
+        height: calc(${controlHeight} - ${smallPadding});
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply packages/nimble-components/src/select/styles.ts]: Yep, your `border-box` change fixed this.
